File: python/day12/day12.py
from collections import defaultdict

input = open('input').read()
grid = [list(it) for it in input.split('\n')]

WIDTH = len(grid[0])
HEIGHT = len(grid)

sx, sy = 0, 0
ex, ey = 0, 0

directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

for y, row in enumerate(grid):
    for x, char in enumerate(row):
        if char == 'S':
            sx, sy = x, y
        if char == 'E':
            ex, ey = x, y

grid[sy][sx] = 'a'
grid[ey][ex] = 'z'

def find_shortest_path(sx, sy, reverse = False):
    paths = defaultdict(lambda: 10e15)
    paths[(sy, sx)] = 0
    should_search = [(sx, sy, 0)]
    while len(should_search) > 0:
        x, y, steps = should_search[0]
        should_search = should_search[1:]
        from_height = ord(grid[y][x])

        for (dx, dy) in directions:
            nx = x + dx
            ny = y + dy
            if nx < 0 or nx >= WIDTH or ny < 0 or ny >= HEIGHT:
                continue
                
            to_height = ord(grid[ny][nx])
            
            # If we're searching for a path up we can move down as much as we want, but only up by 1 level at a time
            if not reverse and not from_height - to_height >= -1:
                continue
            
            # If we're searching in reverse we can only go down one level but up as much as we want
            if reverse and not from_height - to_height <= 1:
                continue

            if paths[(ny, nx)] > steps + 1:
                paths[(ny, nx)] = steps + 1
                should_search.append((nx, ny, steps + 1))
    return paths

part1 = find_shortest_path(sx, sy)[ey, ex]
print(part1)

part2 = min([dist for ((y, x), dist) in find_shortest_path(ex, ey, True).items() if grid[y][x] == 'a'])
print(part2)

File: php/day12/day12.php
<?php

$input = file_get_contents('input');
$grid = array_map(str_split(...), explode(PHP_EOL, $input));


$sx = 0;
$sy = 0;

$ex = 0;
$ey = 0;


foreach ($grid as $y => $row) {
    foreach ($row as $x => $char) {
        if ($char == 'S') {
            $sx = $x;
            $sy = $y;
        } else if ($char == 'E') {
            $ex = $x;
            $ey = $y;
        }
    }
}

$grid[$sy][$sx] = 'a';
$grid[$ey][$ex] = 'z';

function find_shortest_path($grid, int $sx, int $sy, $reverse = false) {
    $width = count($grid[0]);
    $height = count($grid);
    $directions = [[0,1],[0,-1],[-1,0],[1,0]];
    $paths = [];
    $paths["$sx.$sy"] = 0;
    $queue = [[$sx, $sy, 0]];

    while (count($queue) > 0) {
        [$x, $y, $distance] = array_shift($queue);
        $from_height = ord($grid[$y][$x]);
        
        foreach ($directions as $direction) {
            [$dy, $dx] = $direction;
            $nx = $x + $dx;
            $ny = $y + $dy;
            if ($nx < 0 || $nx >= $width ||  $ny < 0 || $ny >= $height) {
                continue;
            }

            $to_height = ord($grid[$ny][$nx] ?? throw new RuntimeException("$nx $ny"));

            if ($reverse == false && false === ($from_height - $to_height >= -1)) {
                continue;
            }

            if ($reverse == true && false === ($from_height - $to_height) <= 1) {
                continue;
            }

            $current_best = $paths["$nx.$ny"] ?? 99999999999;
            if ($current_best > $distance + 1) {
                $paths["$nx.$ny"] = $distance + 1;
                $queue[] = [$nx, $ny, $distance + 1];
            }
        }
    }
    
    return $paths;
}


echo find_shortest_path($grid, $sx, $sy)["$ex.$ey"] . PHP_EOL;

$best = 9871293948728;
foreach (find_shortest_path($grid, $ex, $ey, true) as $key => $value) {
    [$x, $y] = explode('.', $key);
    if ($grid[$y][$x] == 'a' && $value < $best) {
        $best = $value;
    }
}

echo $best . PHP_EOL;